### PR TITLE
Add support for replace/original to C# parser; add SourceGenerator tests; incorporate PR feedback

### PR DESCRIPTION
**Pull request by cston**
Add support for replace/original to C# parser; add tests; and incorporate PR feedback from 56136680f52ea022a4d93dc7bb5fe69ea1c5dbdf.

@dotnet/roslyn-compiler, @dotnet/roslyn-ide, @CyrusNajmabadi, @mattwar please review

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -1462,6 +1418,63 @@ private void ParseModifiers(SyntaxListBuilder tokens)
             }
         }
 
+        /// <summary>
+        /// Return true if the current contextual keyword should be treated as a modifier.
+        /// </summary>
+        private bool IsContextualKeywordActualModifier()
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I feel like the name would be better if it matched your comment.  i.e. "CurrentContextualKeywordShoudlBeTreatedAsAModifier()"

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 2]
```diff
@@ -1168,7 +1168,7 @@ internal override bool CanAddProjectReference(ProjectId referencingProject, Proj
         /// A trivial implementation of <see cref="IVisualStudioWorkspaceHost" /> that just
         /// forwards the calls down to the underlying Workspace.
         /// </summary>
-        private sealed class VisualStudioWorkspaceHost : IVisualStudioWorkspaceHost, IVisualStudioWorkingFolder
+        protected class VisualStudioWorkspaceHost : IVisualStudioWorkspaceHost, IVisualStudioWorkingFolder
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: why?

Comment (cston) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: This is reverting a possibly incorrect change that you noticed in the previous PR.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 3]
```diff
@@ -376,7 +372,30 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
...
+                if (!sourceGenerators.IsEmpty)
+                {
+                    // TODO: Generated source path should include a "GeneratedFiles" subdirectory
+                    // so that "build clean" can be modified to delete the entire directory.
+                    var trees = compilation.GenerateSource(sourceGenerators, this.Arguments.OutputDirectory, writeToDisk: true, cancellationToken: cancellationToken);
+                    if (!trees.IsEmpty)
+                    {
+                        compilation = compilation.AddSyntaxTrees(trees);
+                        // Are there parse warnings? If so, we're probably reporting warnings
+                        // from the previous GetParseDiagnostics() call again here.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: Does this need a follow-up? Do we have an issue tracking this work? I suppose it could be a TODO comment without an issue instead.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Renamed `ShouldCurrentContextualKeywordBeTreatedAsModifier`.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: Changed comment to a TODO.

**Comment by gafter** [comment]
In LanguageParser.cs, reset points do not appear to save and restore `_syntaxFactoryContext.IsInReplace`

**Comment by AlekseyTs** [comment]
Also should investigate if adding /removing the ```replace``` modifier should be specially tracked by an incremental parser.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 4]
```diff
@@ -643,6 +643,21 @@
       <summary>Creates an BaseExpressionSyntax node.</summary>
     </FactoryComment>
   </Node>
+  <Node Name="OriginalExpressionSyntax" Base="InstanceExpressionSyntax">
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: An alternate implementation would be to just have `original` be an ordinary identifier token, like `value` in a property setter, and let semantic analysis take care of it (i.e. bind it to a method).

Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: @gafter I'll keep that approach in mind. For now, this follows the approach used for `AwaitExpressionSyntax`.

**Comment by cston** [comment]
@gafter, @alekseyts Thanks for identifying the incremental parsing issues. I've updated `GetResetPoint()`  and `Reset()` and added some incremental parsing tests. For now, those tests are failing.

**Comment by cston** [comment]
@gafter, @alekseyts, @vsadov Updated PR to include binding and metadata writing of replaced members. Replaced members are now represented in the containing type similar to partial methods - the replacing member is included in `GetMembers()` and the replaced member is a property on the replacing member. Note that a number of the binding tests for `replace ... original` fail currently.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs [hunk 5]
```diff
@@ -513,6 +513,20 @@ private Symbol GetMemberSymbol(string memberName, TextSpan memberSpan, NamedType
                                 return implementation;
                             }
                         }
+
+                        var method = (MethodSymbol)sym;
+                        while (true)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs]: Consider adding a comment explaining why do we need the loop.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 6]
```diff
@@ -1637,6 +1639,70 @@ private BoundExpression BindCastCore(ExpressionSyntax node, BoundExpression oper
...
+            var receiver = symbol.IsStatic ? null : ThisReference(syntax, symbol.ContainingType, wasCompilerGenerated: true);
+            var replaced = symbol.Replaced;
+            Debug.Assert((object)replaced != null);
+            Debug.Assert(replaced.Kind == symbol.Kind);
+
+            switch (symbol.Kind)
+            {
+                case SymbolKind.Method:
+                    {
+                        var replacedMethod = (MethodSymbol)replaced;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: In case of generic method, I think we should construct it with the current method type arguments/parameters.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 7]
```diff
@@ -1637,6 +1639,70 @@ private BoundExpression BindCastCore(ExpressionSyntax node, BoundExpression oper
...
+                        return new BoundMethodGroup(
+                            syntax,
+                            default(ImmutableArray<TypeSymbol>),
+                            replacedMethod.Name,
+                            ImmutableArray.Create(replacedMethod),
+                            lookupSymbolOpt: replacedMethod,
+                            lookupError: null,
+                            flags: BoundMethodGroupFlags.None,
+                            receiverOpt: receiver,
+                            resultKind: LookupResultKind.Viable);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Do we want subsequent access to the replaced method to be virtual or non-virtual?

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 8]
```diff
@@ -4227,9 +4227,6 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ElementIsExpected" xml:space="preserve">
     <value>element is expected</value>
   </data>
-  <data name="MustCallSetMethodTestData" xml:space="preserve">
-    <value>Must call SetMethodTestData(ConcurrentDictionary) before calling SetMethodTestData(MethodSymbol, ILBuilder)</value>
-  </data>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Is this change feature related?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs [hunk 9]
```diff
@@ -326,6 +326,31 @@ public sealed override bool IsVirtual
             get { return (_modifiers & DeclarationModifiers.Virtual) != 0; }
         }
 
+        internal sealed override bool IsReplace
+        {
+            get { return (_modifiers & DeclarationModifiers.Replace) != 0; }
+        }
+
+        private SourceEventSymbol _replacedBy;
+        private SourceEventSymbol _replaced;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: Please declare all fields at the top of the type declaration, before all the constructors.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 7 above)
Comment (mattwar) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: accessing the member via replace should not be a virtual access since it should always access exactly the replaced member, not an overload.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 10]
```diff
@@ -669,6 +675,36 @@ internal sealed override OverriddenOrHiddenMembersResult OverriddenOrHiddenMembe
...
+            get
+            {
+                return (object)_replacedBy == null ?
+                    base.MetadataName :
+                    GeneratedNames.ReplacedMemberName(Name, uniqueId: 0);
+            }
+        }
+
+        private SourceMethodSymbol _replacedBy;
+        private SourceMethodSymbol _replaced;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: Please move to the top of type declaration.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs [hunk 11]
```diff
@@ -574,6 +574,31 @@ internal bool IsNew
             get { return (_modifiers & DeclarationModifiers.New) != 0; }
         }
 
+        internal sealed override bool IsReplace
+        {
+            get { return (_modifiers & DeclarationModifiers.Replace) != 0; }
+        }
+
+        private SourcePropertySymbol _replacedBy;
+        private SourcePropertySymbol _replaced;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs]: Please move to the top of type declaration.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs [hunk 12]
```diff
@@ -574,6 +574,31 @@ internal bool IsNew
             get { return (_modifiers & DeclarationModifiers.New) != 0; }
         }
 
+        internal sealed override bool IsReplace
+        {
+            get { return (_modifiers & DeclarationModifiers.Replace) != 0; }
+        }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs]: Should we override MetadataName?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs]: And for events?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 7 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Then we should probably make sure SuppressVirtualCalls is true for the receiver.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 8 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: The changes around `CompilationTestData` are not specific to this feature but are needed here to allow multiple methods with the same signature.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs (hunk 5 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs]: Added.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 7 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: The replaced method is no longer virtual. Regardless, the call doesn't need to be virtual so emit now generates `Call` rather than `CallVirt` for replaced methods.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs (hunk 9 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: Moved.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs (hunk 12 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs]: Replaced properties and events are not emitted to metadata, only the accessors are.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 7 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: IOperation pulls this information from BoundTree, so, if the symbol is virtual, it will likely to treat this call as virtual.

**Comment by cston** [comment]
All feedback has been addressed in most recent commits.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 7 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Updated `IInvocationExpression` implementation to return false for `IsVirtual` for replaced method calls.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 13]
```diff
@@ -42,7 +39,10 @@ internal partial class BoundCall : IInvocationExpression
 
         IOperation IInvocationExpression.Instance => this.Method.IsStatic ? null : this.ReceiverOpt;
 
-        bool IInvocationExpression.IsVirtual => (this.Method.IsVirtual || this.Method.IsAbstract || this.Method.IsOverride) && !this.ReceiverOpt.SuppressVirtualCalls;
+        bool IInvocationExpression.IsVirtual =>
+            (this.Method.IsVirtual || this.Method.IsAbstract || this.Method.IsOverride) &&
+            ((object)this.Method.ReplacedBy == null) &&
+            !this.ReceiverOpt.SuppressVirtualCalls;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: I think there are two more similar places in this file.

Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Thanks. Fixed.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 6 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Good catch. Fixed.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: @cston Are you updating this PR with the fixes?

Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Yes. For this particular issue, the fix is to pass `replacedByMethod.TypeArguments` at line 1665.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 14]
```diff
@@ -437,6 +438,12 @@ private void CompileNamedType(NamedTypeSymbol containingType)
                                 hasStaticConstructor = true;
                             }
 
+                            // Replaced methods are not included in GetMembers().
+                            foreach (var replaced in method.GetReplacedMethods())
+                            {
+                                CompileMethod(replaced, extendedOrdinal, ref processedInitializers, synthesizedSubmissionFields, compilationState);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: The code gen might be more stable (e.g. for edit-and-continue scenarios) if the ordinal were per method rather than for the whole enclosing type.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: Wait... how is this ordinal used?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 15]
```diff
@@ -1462,6 +1418,60 @@ private void ParseModifiers(SyntaxListBuilder tokens)
...
+            }
+
+            bool isModifier = false;
+
+            // Some of our helpers start at the current token, so we'll have to advance for their
+            // sake and then backtrack when we're done.  Don't leave this block without releasing
+            // the reset point.
+            ResetPoint resetPoint = GetResetPoint();
+
+            this.EatToken(); //move past "async"
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: "async" or "replace"

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 16]
```diff
@@ -1451,70 +1455,31 @@ private void CheckMemberNameConflicts(DiagnosticBag diagnostics)
...
-                    var method = symbol as SourceMethodSymbol;
+                    // Does this conversion collide with any previously-seen conversion?
                     if ((object)conversion != null)
                     {
-                        // Does this conversion collide *as a conversion* with any previously-seen
-                        // conversion?
-
-                        SourceUserDefinedConversionSymbol previousConversion;
-                        if (conversionsAsConversions.TryGetValue(conversion, out previousConversion))
+                        if (!conversionsBySignature.Add(conversion))
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Was the `conversionsAsConversions` stuff unnecessary?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 17]
```diff
@@ -1451,70 +1455,31 @@ private void CheckMemberNameConflicts(DiagnosticBag diagnostics)
                     // second and third categories as follows:
 
                     var conversion = symbol as SourceUserDefinedConversionSymbol;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: (I know you didn't write this line of code but) is there a test that can accomplish this semantically without depending on the specific implementation hierarchy?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 18]
```diff
@@ -1451,70 +1455,31 @@ private void CheckMemberNameConflicts(DiagnosticBag diagnostics)
...
-                        {
-                            ReportMethodSignatureCollision(diagnostics, method, previousMethod);
-                        }
-                        else
-                        {
-                            // We haven't seen this method before. Make a note of it in case
-                            // we see a colliding method later.
-                            methodsBySignature[method] = method;
-                        }
+                foreach (var group in methodsBySignature.Values)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: I don't understand why you separated the detection of the conflict (by producing multiple values into a `OneOrMany`) from the reporting. This appears to have the same behavior as before but at the cost of storing more transient data.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 19]
```diff
@@ -2326,26 +2289,105 @@ internal Binder GetBinder(CSharpSyntaxNode syntaxNode)
...
+                        continue;
+                    }
+
+                    Symbol last = null;
+                    foreach (var member in group)
+                    {
+                        var method = member as SourceMethodSymbol;
+                        if ((object)method != null && method.IsPartial)
+                        {
+                            continue; // PROTOTYPE: Allow replacing partial methods.
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: All the `PROTOTYPE` comments should include `PROTOTYPE(generators)`, please.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs (hunk 14 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: Added a comment to `extendedOrdinal` that it is used to ensure replaced methods have unique ids.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 20]
```diff
@@ -2411,15 +2453,19 @@ private static ImmutableArray<Symbol> FixPartialMember(ImmutableArray<Symbol> sy
             SourceMemberMethodSymbol.InitializePartialMethodParts(definition, implementation);
 
             // a partial method is represented in the member list by its definition part:
+            return Remove(symbols, implementation);
+        }
+
+        private static ImmutableArray<Symbol> Remove(ImmutableArray<Symbol> symbols, Symbol symbol)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Given that replace/original may be used more than partial methods, doing this over and over again can be inefficient (i.e. quadratic). Does it make sense to `RemoveAll` a set of symbols rather than one at a time? (a PROTOTYPE comment or git issue would be a good idea to investigate this later).

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 21]
```diff
@@ -669,6 +678,33 @@ internal sealed override OverriddenOrHiddenMembersResult OverriddenOrHiddenMembe
             }
         }
 
+        public override string MetadataName
+        {
+            get
+            {
+                return (object)_replacedBy == null ?
+                    base.MetadataName :
+                    GeneratedNames.ReplacedMemberName(Name, uniqueId: 0);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: I'm a bit confused about how the replaced and originals are linked together. Are they like a linked list, where each replace method links to and from the method it replaces, or does the original method have a list of all of the replacements, or both? Is it worth adding a comment summarizing the strategy somewhere?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 19 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Even though `PROTOTYPE` comments should be removed before merging into the future branch?

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 22]
```diff
@@ -1609,21 +1612,14 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
             CompilationTestData testData,
             CancellationToken cancellationToken);
 
-        internal EmitResult Emit(
-            EmitStreamProvider peStreamProvider,
-            EmitStreamProvider pdbStreamProvider,
-            EmitStreamProvider xmlDocumentationStreamProvider,
-            EmitStreamProvider win32ResourcesStreamProvider,
+        internal CommonPEModuleBuilder PreCompile(
```
Comment (gafter) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: It would be nice to have some documentation on the relative role of PreCompile, Compile, and PostCompile. Just a sentence of two on each, please.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 19 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Yes. It is likely we'll integrate early to facilitate testing multiple featured together by end users before every eye is crossed and tee dotted.

**Comment by gafter** [comment]
LGTM aside from the few comments I made.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 15 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Changed to `contextual keyword`.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 19 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Changed all to ```PROTOTYPE(generators)```.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs (hunk 21 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: The `Replaced` and `ReplacedBy` members reference each other. The `ReplacedBy` member is included in ```GetMembers()``` on the containing type while the `Replaced` member is only available from `ReplacedBy.Replaced`.

Add comments to `Symbol.Replaced` and `Symbol.ReplacedBy`.

Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: The `Replaced` and `ReplacedBy` members reference each other. The `ReplacedBy` member is included in `GetMembers()` of the containing type while the `Replaced` member is only available from `ReplacedBy.Replaced`.

Added comments to `Symbol.Replaced` and `Symbol.ReplacedBy`.

File: src/Compilers/Core/Portable/Compilation/Compilation.cs (hunk 22 above)
Comment (cston) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: Added comments.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 20 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Added a `PROTOTYPE` comment. Note this would only allow removing all replace symbols at once that have the same name.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 17 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Perhaps this can be accomplished by comparing ```symbol.Kind == SymbolKind.Method && (MethodSymbol)symbol).MethodKind == MethodKind.Conversion``` and then declaring `conversionsBySignature` as a dictionary of `MethodSymbol`. Will leave as is for now though.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 18 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Good catch, thanks. This was left over from an earlier attempt to match up replaced methods. I've reverted the change.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 16 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: I believe method signature collisions can be handled by `methodsBySignature` alone. As a result, for consistency, `conversionsAsConversions` was renamed to `conversionsBySignature`.
